Firestore/core/src/firebase/firestore/api/firestore.mm

```diff
@@ -143,6 +144,25 @@
   [client_ shutdownWithCallback:std::move(callback)];
 }
 
+void Firestore::ClearPersistence(util::StatusCallback callback) {
+  worker_queue()->Enqueue([this, callback] {
+    std::lock_guard<std::mutex> lock{mutex_};
```

Comment:
In general favor code that holds locks for the least amount of time. This is holding the lock for the entirety of the lambda, which includes the I/O to actually perform the deletion. This seems undesirable.

Note that because we're already on the async queue, we're already guaranteed mutual exclusion with any other operation on the async queue so here we don't really need to prevent the user from calling `EnsureClientConfigured()` and even if they do so to call something other than `Shutdown()` the queue ensures that the effect of that call can't happen until after our work in `ClearPersistence` is done.

A way to rephrase this that makes the lock scope smaller would be to make the lambda

```c++
  worker_queue()->Enqueue([this, callback] {
    auto Yield = [callback](Status status) {
      if (callback) {
        this->user_executor_->Execute([=] { callback(status); });
      }
    }

    {
      std::lock_guard<std::mutex> lock{mutex_};
      if (client_ && !client().isShutdown) {
        Yield(util::Status(
            FirestoreErrorCode::FailedPrecondition,
            "Persistence cannot be cleared while the client is running."));
        return;
      }
    }

    DatabaseInfo database_info(database_id_, persistence_key_,
                               settings_.host(), settings_.ssl_enabled());
    Yield([FSTLevelDB clearPersistence:database_info]);
  });